### PR TITLE
Add Parse func accepting various types

### PR DESCRIPTION
**Pull request by lgierth**
@Kubuxu @whyrusleeping @keks for review

### CODE REVIEW

**Comment by lgierth** [comment]
For explanation, this func is useful in Core API, where we want to be flexible in what the we expect as function arguments. Right now we achieve this flexibility through requiring strings, but that's less than ideal as it involved lots of casting from-and-to. With this change, the Core API functions can just accept `interface{}`, and let the cid package take care of turning that into a Cid object.

File: cid.go [hunk 1]
```diff
@@ -47,6 +48,25 @@ type Cid struct {
 	hash    mh.Multihash
 }
 
+func Parse(v interface{}) (*Cid, error) {
+	switch v.(type) {
+	case string:
+		v2 := v.(string)
+		if strings.Contains(v2, "/ipfs/") {
+			return Decode(strings.Split(v2, "/ipfs/")[1])
```
Comment (lgierth) [inline cid.go]: Note that this also accepts https://ipfs.io/ipfs/hash :)

**Comment by keks** [comment]
You *could* check the errors returned by `Parse`. If tests pass there is no difference, but if they don't the error value gives you guidance what the problem is.

Else LGTM.

File: cid.go [hunk 2]
```diff
@@ -47,6 +48,25 @@ type Cid struct {
...
+		}
+		return Decode(v2)
+	case []byte:
+		return Cast(v.([]byte))
+	case mh.Multihash:
+		return NewCidV0(v.(mh.Multihash)), nil
+	case *Cid:
+		return v.(*Cid), nil
+	default:
+		return nil, fmt.Errorf("can't parse %+v as Cid", v)
```
Comment (Kubuxu) [inline cid.go]: but it might be reference to struct name

Comment (lgierth) [inline cid.go]: Yep that's referencing the struct we're trying to get out of this

**Comment by lgierth** [comment]
> You could check the errors returned by Parse. If tests pass there is no difference, but if they don't the error value gives you guidance what the problem is.

Yep updated, good point

File: cid.go [hunk 3]
```diff
@@ -47,6 +48,25 @@ type Cid struct {
 	hash    mh.Multihash
 }
 
+func Parse(v interface{}) (*Cid, error) {
+	switch v.(type) {
```
Comment (lgierth) [inline cid.go]: > Instead of casting everywhere you can do switch with assignment, switch value := v.(type) {.

Updated

File: cid.go [hunk 4]
```diff
@@ -47,6 +48,25 @@ type Cid struct {
 	hash    mh.Multihash
 }
 
+func Parse(v interface{}) (*Cid, error) {
```
Comment (lgierth) [inline cid.go]: > it can give imprecise result

I don't follow, have I overlooked something?

**Comment by lgierth** [comment]
Also DRY'ed up the tests

File: cid.go (hunk 4 above)
Comment (Kubuxu) [inline cid.go]: Nah, doesn't matter.

**Pull request merged**